common/unsafe/src/main/java/org/apache/spark/unsafe/UTF8StringBuilder.java

```diff
@@ -93,7 +93,38 @@ public void appendBytes(Object base, long offset, int length) {
     cursor += length;
   }
 
+  public void appendByte(byte singleByte) {
+    grow(1);
+    buffer[cursor - Platform.BYTE_ARRAY_OFFSET] = singleByte;
+    cursor++;
+  }
+
   public UTF8String build() {
     return UTF8String.fromBytes(buffer, 0, totalSize());
   }
+
+  public void appendCodePoint(int codePoint) {
+    if (codePoint <= 0x7F) {
+      appendByte((byte) codePoint);
+    }
+    else if (codePoint <= 0x7FF) {
+      appendByte((byte) (0xC0 | (codePoint >> 6)));
+      appendByte((byte) (0x80 | (codePoint & 0x3F)));
+    }
+    else if (codePoint <= 0xFFFF) {
+      appendByte((byte) (0xE0 | (codePoint >> 12)));
+      appendByte((byte) (0x80 | ((codePoint >> 6) & 0x3F)));
+      appendByte((byte) (0x80 | (codePoint & 0x3F)));
+    }
+    else if (codePoint <= 0x10FFFF) {
+      appendByte((byte) (0xF0 | (codePoint >> 18)));
+      appendByte((byte) (0x80 | ((codePoint >> 12) & 0x3F)));
+      appendByte((byte) (0x80 | ((codePoint >> 6) & 0x3F)));
+      appendByte((byte) (0x80 | (codePoint & 0x3F)));
+    }
+    else {
+      throw new IllegalArgumentException("Invalid Unicode codePoint: " + codePoint);
+    }
```

Comment:
```suggestion
    if (codePoint <= 0x7F) {
      appendByte((byte) codePoint);
    } else if (codePoint <= 0x7FF) {
      appendByte((byte) (0xC0 | (codePoint >> 6)));
      appendByte((byte) (0x80 | (codePoint & 0x3F)));
    } else if (codePoint <= 0xFFFF) {
      appendByte((byte) (0xE0 | (codePoint >> 12)));
      appendByte((byte) (0x80 | ((codePoint >> 6) & 0x3F)));
      appendByte((byte) (0x80 | (codePoint & 0x3F)));
    } else if (codePoint <= 0x10FFFF) {
      appendByte((byte) (0xF0 | (codePoint >> 18)));
      appendByte((byte) (0x80 | ((codePoint >> 12) & 0x3F)));
      appendByte((byte) (0x80 | ((codePoint >> 6) & 0x3F)));
      appendByte((byte) (0x80 | (codePoint & 0x3F)));
    } else {
      throw new IllegalArgumentException("Invalid Unicode codePoint: " + codePoint);
    }
```
Is there possibly some Java functionality for doing this, or ICU functionality for doing this, so we don't have to do it ourselves here?

Reply:
There are, but for the ones i found, we would have to allocate at least something on the heap. So i went for this approach because of performance.